usaspending_api/etl/management/commands/swap_in_new_table.py

```diff
@@ -0,0 +1,292 @@
+            " It is advised to not allow Foreign Key constraints since they can cause deadlock.",
+        )
+
+    def handle(self, *args, **options):
+        self.db_table_name = options["table"]
+        self.temp_db_table_name = f"{self.db_table_name}_temp"
+
+        with connection.cursor() as cursor:
+            self.validate_state_of_tables(cursor, options)
+            self.swap_index_sql(cursor)
+            self.swap_constraints_sql(cursor)
+            self.swap_table_sql(cursor)
+            if not options["keep_old_data"]:
+                self.drop_old_table_sql(cursor)
+            self.extra_sql(cursor)
+
+    def swap_index_sql(self, cursor):
+        logging.info("Renaming indexes of the new and old tables.")
+        temp_indexes = self.query_result_lookup["temp_table_indexes"]
+        curr_indexes = self.query_result_lookup["curr_table_indexes"]
+        rename_sql = []
+        sql_template = "ALTER INDEX {old_index_name} RENAME TO {new_index_name};"
+        for val in curr_indexes:
+            old_name = val["indexname"]
+            new_name = f"{old_name}_old"
+            rename_sql.append(sql_template.format(old_index_name=old_name, new_index_name=new_name))
+        for val in temp_indexes:
+            old_name = val["indexname"]
+            new_name = re.match("^(.*)_temp$", old_name, flags=re.I)[1]
+            rename_sql.append(sql_template.format(old_index_name=old_name, new_index_name=new_name))
+
+        if rename_sql:
+            cursor.execute("\n".join(rename_sql))
+
+    def swap_constraints_sql(self, cursor):
+        logging.info("Renaming constraints of the new and old tables.")
+        temp_constraints = self.query_result_lookup["temp_table_constraints"]
+        curr_constraints = self.query_result_lookup["curr_table_constraints"]
+        rename_sql = []
+        sql_template = "ALTER TABLE {table_name} RENAME CONSTRAINT {old_constraint_name} TO {new_constraint_name};"
+        for val in curr_constraints:
+            old_name = val["constraint_name"]
+            new_name = f"{old_name}_old"
+            rename_sql.append(
+                sql_template.format(
+                    table_name=self.db_table_name, old_constraint_name=old_name, new_constraint_name=new_name
+                )
+            )
+        for val in temp_constraints:
+            old_name = val["constraint_name"]
+            new_name = re.match("^(.*)_temp$", old_name, flags=re.I)[1]
+            rename_sql.append(
+                sql_template.format(
+                    table_name=self.temp_db_table_name, old_constraint_name=old_name, new_constraint_name=new_name
+                )
+            )
+
+        if rename_sql:
+            cursor.execute("\n".join(rename_sql))
+
+    @transaction.atomic
+    def swap_table_sql(self, cursor):
+        logging.info("Renaming the new and old tables")
+        sql_template = "ALTER TABLE {old_table_name} RENAME TO {new_table_name};"
```

Comment:
I thought we didn't want this in a transaction, because if FKs exist on the table to rename, an ACCESS EXCLUSIVE LOCK is taken on the renamed table, and it cascades to the tables where FKs exist and can cause a deadlock scenario

Reply:
The ACCESS EXCLUSIVE is taken on two tables introducing possibility for deadlock when the FK constraints are renamed or dropped in a transaction because it affects the trigger on the downstream table. Alter the primary table only takes a lock on the single table. I figured having this as the only part in a transaction will remove the possibility of a `relation not found` error for any queries that happen in the milliseconds it takes to run.